### PR TITLE
DNM, build: switch the make check bot to Clang.

### PR DESCRIPTION
**Pull request by rzarzynski**
DNM, currently for testing only.

this PR depends on following changes:

- [x] https://github.com/ceph/ceph/pull/42742
- [ ] https://github.com/axboe/fio/pull/1261
- [ ] bump up the fio version to include https://github.com/axboe/fio/pull/1261

Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by rzarzynski** [comment]
OK, the transition seems to be in effect. `Clang` has been selected on it actually found a code-smell:

```
FAILED: src/test/fio/CMakeFiles/fio_ceph_objectstore.dir/fio_ceph_objectstore.cc.o 
/usr/bin/ccache /usr/bin/clang++-10  -DBOOST_ALL_NO_LIB -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DBOOST_DATE_TIME_DYN_LINK -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_RANDOM_DYN_LINK -DBOOST_SYSTEM_DYN_LINK -DBOOST_THREAD_DYN_LINK -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -Dfio_ceph_objectstore_EXPORTS -Isrc/include -I../src -isystem /opt/ceph/include -isystem include -isystem ../src/xxHash -isystem ../src/rapidjson/include -isystem src/fio -isystem ../src/rocksdb/include -isystem /usr/include/fuse -g -g -fPIC   -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -include /home/jenkins-build/build/workspace/ceph-pull-requests/build/src/fio/config-host.h -std=gnu++17 -std=c++17 -MD -MT src/test/fio/CMakeFiles/fio_ceph_objectstore.dir/fio_ceph_objectstore.cc.o -MF src/test/fio/CMakeFiles/fio_ceph_objectstore.dir/fio_ceph_objectstore.cc.o.d -o src/test/fio/CMakeFiles/fio_ceph_objectstore.dir/fio_ceph_objectstore.cc.o -c ../src/test/fio/fio_ceph_objectstore.cc
In file included from ../src/test/fio/fio_ceph_objectstore.cc:26:
In file included from src/fio/fio.h:18:
In file included from src/fio/thread_options.h:6:
In file included from src/fio/options.h:8:
src/fio/parse.h:128:13: error: arithmetic on a pointer to void
        return ret + offset;
               ~~~ ^
1 error generated.
```

Fixing.

**Comment by rzarzynski** [comment]
Hmm, why do we build `WITH_FIO` for the `make check`?

**Comment by rzarzynski** [comment]
```
97% tests passed, 6 tests failed out of 238

Total Test time (real) = 491.55 sec

The following tests FAILED:
	132 - check-generated.sh (Failed)
	133 - readable.sh (Failed)
	224 - unittest-transaction-manager (Failed)
	227 - unittest-seastore-cache (Failed)
	231 - unittest-seastore (Failed)
	232 - unittest-seastore-randomblock-manager (Failed)
```

**Comment by rzarzynski** [comment]
jenkins retest this please

**Comment by rzarzynski** [comment]
The same tests failed again:

```

97% tests passed, 6 tests failed out of 238

Total Test time (real) = 492.18 sec

The following tests FAILED:
	132 - check-generated.sh (Failed)
	133 - readable.sh (Failed)
	224 - unittest-transaction-manager (Failed)
	227 - unittest-seastore-cache (Failed)
	231 - unittest-seastore (Failed)
	232 - unittest-seastore-randomblock-manager (Failed)
Errors while running CTest
```

**Comment by rzarzynski** [comment]
jenkins retest this please

**Comment by rzarzynski** [comment]
Related: #42607.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by wjwithagen** [comment]
> The same tests failed again:
> 	132 - check-generated.sh (Failed)
> 	133 - readable.sh (Failed)

These 2 test failing with Clang seems rather familiar with my tests on FreeBSD. 

This is the diff I have for check-generated.sh:
```
--- a/src/test/encoding/check-generated.sh
+++ b/src/test/encoding/check-generated.sh
@@ -16,6 +16,20 @@ numtests=0
 echo "checking ceph-dencoder generated test instances..."
 echo "numgen type"
 while read type; do
+    # Skip tests that abort
+    if [    $type = OSDMap -o $type = ServiceMap \
+        -o $type = 'cls::rbd::MirrorImageSiteStatusOnDisk' \
+       ]; then
+       continue
+    fi
+    # Skip tests that fail
+    if [    $type =  'rados::cls::lock::locker_info_t' \
+        -o $type = 'ServiceMap::Daemon' \
+        -o $type = 'entity_addrvec_t' \
+        -o $type = 'entity_inst_t' \
+       ]; then
+       continue
+    fi
     num=`ceph-dencoder type $type count_tests`
     echo "$num $type"
     for n in `seq 1 1 $num 2>/dev/null`; do
```

The reason for skipping these tests is that one way or another it seems that GCC clears more memory, and the structures contain 0's on certain possitions where they do not in case of the Clang compiler.
For the `entity_*` this is sort of expected, since the FreeBSD addr struct is quite different from the Linux one.
Sort of cheating, but figuring this out was quite an effort, and in the end I decided to skip the tests, since I wanted to focus on other things.

In the `readable.sh` more or less the same. you have to dig into the actual difference in output to find out that there are tiny conversion differences. For some of the tests that was fixed by starting to compare the json output.
That way stuffing fields did not matter anymore, which was the problem in binary compares.
I think it was Sage that fixed this part, after which my FreeBSD test did not have problems any longer.
But perhaps on Linux there are still some fields left that differ.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins retest this please

**Comment by tchaikov** [comment]
```
The following tests FAILED:
	225 - unittest-transaction-manager (Failed)
	228 - unittest-seastore-cache (Failed)
	232 - unittest-seastore (Failed)
	233 - unittest-seastore-randomblock-manager (Failed)
```

when accessing `interrupt_cond<InterruptCond>` at

https://github.com/ceph/ceph/blob/0ff2e28188997853804c97d42dd070c38127c94d/src/crimson/common/interruptible_future.h#L102

the value of `%fs` segment register is 0. and the offset is also 0, so we have segfault, and ASan errors out.

**Comment by rzarzynski** [comment]
@tchaikov: aren't we just missing the definition of this `thread_local` variable template?

https://github.com/ceph/ceph/blob/0ff2e28188997853804c97d42dd070c38127c94d/src/crimson/common/interruptible_future.h#L59-L60

Prepending with `inline` would do the trick with ODR.

**Comment by rzarzynski** [comment]
Huh, we could check with the `make check` as this env. is already able to build the stuff!

**Comment by rzarzynski** [comment]
Nah, we still miss something:

```
The following tests FAILED:
	225 - unittest-transaction-manager (Failed)
	228 - unittest-seastore-cache (Failed)
	232 - unittest-seastore (Failed)
	233 - unittest-seastore-randomblock-manager (Failed)
```

**Comment by tchaikov** [comment]
@rzarzynski #42715

File: src/script/run-make.sh [hunk 1]
```diff
@@ -73,6 +74,7 @@ function prepare() {
     elif type apt-get > /dev/null 2>&1 ; then
         install_cmd="apt-get install -y"
         which_pkg="debianutils"
+        compiler_pkg="clang-10"
```
Comment (tchaikov) [inline src/script/run-make.sh:77]: why? i think "clang" depends on "clang-10", and "clang" is more future-proof. on debian sid, "clang" depends on "clang-11", on ubuntu hirsute, "clang" depends on "clang-12".

File: run-make-check.sh [hunk 2]
```diff
@@ -70,8 +70,17 @@ function main() {
         return 1
     fi
     FOR_MAKE_CHECK=1 prepare
+    cxx_compiler=g++
+    c_compiler=gcc
+    for i in $(seq 11 -1 10); do
```
Comment (tchaikov) [inline run-make-check.sh:75]: i s/11/14/ to be more future-proof. the latest version of clang is 12. see https://releases.llvm.org/download.html. and 13 is already packaged by some distros.

File: run-make-check.sh [hunk 3]
```diff
@@ -80,7 +80,7 @@ function main() {
         fi
     done
     # Init defaults after deps are installed.
-    local cmake_opts=" -DCMAKE_CXX_COMPILER=$cxx_compiler -DCMAKE_C_COMPILER=$c_compiler -DWITH_GTEST_PARALLEL=ON -DWITH_FIO=ON -DWITH_CEPHFS_SHELL=ON -DWITH_SPDK=ON -DENABLE_GIT_VERSION=OFF"
+    local cmake_opts=" -DCMAKE_CXX_COMPILER=$cxx_compiler -DCMAKE_C_COMPILER=$c_compiler -DWITH_GTEST_PARALLEL=ON -DWITH_CEPHFS_SHELL=ON -DWITH_SPDK=ON -DENABLE_GIT_VERSION=OFF"
```
Comment (tchaikov) [inline run-make-check.sh:83]: this change practically reverts part of #20535 .


see https://github.com/ceph/ceph/pull/20535/files#diff-54793ec7325559f136d18596b00188c00c48dbb1c253a20cb327b80092201edeR71

i'd suggest just bump its version to the latest release.

**Comment by tchaikov** [comment]
@rzarzynski i incorporated some commits in this changeset into #42715 . could you help review it?

File: run-make-check.sh [hunk 4]
```diff
@@ -70,8 +70,17 @@ function main() {
         return 1
     fi
     FOR_MAKE_CHECK=1 prepare
+    cxx_compiler=g++
+    c_compiler=gcc
+    for i in $(seq 11 -1 14); do
```
Comment (tchaikov) [inline run-make-check.sh:75]: should be `seq 14 -1 11`.

**Comment by rzarzynski** [comment]
jenkins test crimson perf

**Comment by rzarzynski** [comment]
jenkins test crimson perf

**Comment by rzarzynski** [comment]
jenkins test crimson perf

**Comment by rzarzynski** [comment]
jenkins test crimson perf

**Comment by rzarzynski** [comment]
jenkins test crimson perf

**Comment by rzarzynski** [comment]
jenkins test crimson perf

**Comment by rzarzynski** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
@rzarzynski i included the proposed changes in #42777, could you help review it?

**Comment by tchaikov** [comment]
closing in favor of #42777

**Pull request closed without merging**